feat(25.04): rust 1.84 for 25.04

lczyk opened the pull request:
This PR adds the rust rock for 25.04. It is a culmination of much work in chisel releases (see below). The tests here broadly follow the tests for individual slices in chisel releases.

~~[#633](https://github.com/canonical/chisel-releases/pull/633) - ROCKS-2270/libc6-dev-for-25.04~~ *(merged)*
~~└ [#644](https://github.com/canonical/chisel-releases/pull/644) - ROCKS-2270/binutils-for-25.04~~ *(merged)*
~~....└ [#648](https://github.com/canonical/chisel-releases/pull/648) - ROCKS-2270/cpp-for-25.04~~ *(merged)*
........└ [#649](https://github.com/canonical/chisel-releases/pull/649) - ROCKS-2270/gcc-14-for-25.04
............└ [#610](https://github.com/canonical/chisel-releases/pull/610) - ROCKS-2271/rust-1.84-25.04
................└ [#634](https://github.com/canonical/chisel-releases/pull/634) - ROCKS-2271/cargo-1.84-25.04
....................└ [#3](https://github.com/canonical/rust-rock/pull/3) - ROCKS-2271/rust-rock-25.04 **(this PR)**

For now the rockcraft.yaml uses the `lczyk/chisel-releases/ROCKS-2271/cargo-1.84-25.04` branch. This will be updated to the main `chisel-releases` branch once the PRs above are reviewed and merged.

lczyk committed:
chore: update codeowners

lczyk committed:
docs: update readme

lczyk committed:
feat!: update rock name and version

lczyk committed:
feat(testing): setup testing

lczyk committed:
feat: rust sdk rock definition

lczyk committed:
feat(testing): test and version

lczyk committed:
feat(testing): refactor defer

lczyk committed:
feat(testing): gcc hello world

lczyk committed:
feat(testing): rustc hello world

lczyk committed:
feat(testing): cargo hello world

lczyk committed:
fix(testing): slice names on x86

lczyk committed:
chore: appease yamllint

lczyk committed:
feat(testing): test building sudo

lczyk committed:
feat(testing): test building eza

lczyk committed:
feat(testing): test building cargo

lczyk committed:
fix(testing): make run.sh executable

lczyk committed:
feat(testing): test building fd

lczyk committed:
feat(testing): download eza tests from source release

lczyk committed:
feat: cargo test eza

lczyk committed:
feat: cargo test fd

lczyk committed:
feat: test building oxidizr

lczyk committed:
feat: cargo test sudo

lczyk committed:
fix(testing): fd test fix

lczyk committed:
feat: update to use apt-get-mini

lczyk committed:
feat: defer self-test

lczyk committed:
feat: use sponge to prevent races in command execution

lczyk committed:
chore: newlines

lczyk committed:
feat: apt-get mini

lczyk committed:
fix: add --bootstrap to cargo cargo test

lczyk committed:
feat: deb-security-manifest

lczyk committed:
feat: make sure security manifest runs last

lczyk committed:
fix: deb-security-manifest

lczyk committed:
feat: update slice name

lczyk committed:
docs: update description

lczyk committed:
feat: SPDX license format

lczyk committed:
feat: comment out the license expression

lczyk committed:
feat: update the spread backend

lczyk committed:
feat: remove runuser

lczyk committed:
revert: change the spread backend back to 24.04

alesancor1 commented on rust-rock/1.84-25.04/spread/general/test/defer.sh:
This comments applies to the testing in general:

- I am not against having `task.yaml` along with `sh` files, `sh` format will provide the formatting and linting as well as the isolation to be run separately with native tools (no need for spread)
- I'd say the files `run.sh` and  `setup.sh` are more developer-oriented since they just provide a pretty way for the user to run all the tests and see the results, this is essentially an alternative to spread. I don't think they should be committed here because:
  - Spread is an external tool that run the `task.yaml`, so the behaviour here should be the same, an "external" tool that runs the `sh` files
  - The user can choose to run those files in different ways (with spread, separated, or with a custom script).
- `defer.sh` is an utility function that may come handy when shared across all the scripts. If it is needed, place it in  `lib` directory or similar. Though, this file should not be part of the upstream template.
- The folder layout should be:
   ```
     spread/general
          L test1/
                L task.yaml
                L task.sh
          L test2/
                L task.yaml
                L task.sh
         ...
    ```

alesancor1 commented on rust-rock/1.84-25.04/rust_override_build.sh:
Not against it but not convinced either. I think the logic in here is simple enough to be placed inside the `yaml`, especially considering that the slice installation is a temporary thing and will be done via `stage-packages` when slices are merged.

lczyk replied:
yes, but even then there will be the additional symlink making step + copying over the apt-shim. also, exactly the same argument as for tests: syntax highlighting + linting in actual bash files.

lczyk replied:
oh, and one more: like that we get actual bash as opposed to dash

alesancor1 commented on rust-rock/1.84-25.04/apt-shim.sh:
Where is this being used?

lczyk replied:
in `rust_override_build.sh`

alesancor1 replied:
Ok, in any case, why do we need `apt` in the rock?

lczyk replied:
in case someone wants to compile something which needs additional deps. for example sudo needs libpam (see the sudo test)

alesancor1 replied:
I think this is a similar case to the `jdk` rock, where we have one for development and one for apps (I believe), ask James (can't tag him as this repo is not yet onboarded to the repo-automation it seems)

lczyk replied:
the rust rock for apps is not really a thing though. rust is a compiled language. once you compile a binary you don't need the compiler and there is no runtime. (aka this is, by default, a rock for development)

lczyk replied:
> ask James

@zhijie-yang

alesancor1 reviewed (commented):
Overall it looks good, just a few comments regarding testing, please find them below.

Functionality wise, I'd say this is ready to go into `edge`

lczyk committed:
feat: ignore SIGPIPE

lczyk committed:
feat: spellchecker ignore

lczyk committed:
feat: move env

lczyk committed:
feat: refactor tests

lczyk committed:
feat: add common.sh

lczyk committed:
feat: ignore sigpipe in common tests

lczyk committed:
feat: fix wrong declare name

lczyk committed:
feat: go back to eux

alesancor1 commented on rust-rock/1.84-25.04/spread/tmp/.gitkeep:
Do we need this file committed? We can create `tmp` on the fly and `.gitignore` it right? I don't think temp files should be pushed

lczyk committed:
feat: change to apt-bootstrap

lczyk committed:
feat: rm tmp

lczyk commented:
also confirmed with @cjdcordeiro that yes, we can PR an **edge** rock (built from my own repos, rather than merged slices) and then update once slices are merged

lczyk committed:
feat: change to big apt

lczyk committed:
fix: clean refs to apt-bootstrap

lczyk commented:
based on @zhijie-yang's suggestions i've removed apt-bootstrap and changed it to a normal apt slice. i've also explicitly included coreutils since this is an SDK rock

lczyk committed:
feat: refactor tests

lczyk committed:
fix: apt update

lczyk committed:
feat: use launch_container in apt test

lczyk committed:
fix: unbound

lczyk committed:
feat: defer 1.0.3

lczyk committed:
chore: trigger rebuild

lczyk committed:
fix: launch_container

lczyk committed:
feat: base-files bits
